File: frontend/src/app/core/services/reserva.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';
import { ApiResponse, PagedResponse } from '../models/api-response.model';

export interface Reserva {
  id: number;
  codigo: string;
  clienteId: number;
  proveedorId: number;
  servicioId: number;
  mascotaId: number;
  estadoReservaId: number;
  registroVacunacionId?: number;
  modalidadEntrega?: string;
  latitud?: number;
  longitud?: number;
  direccionReferencia?: string;
  fechaReserva: string;
  fechaInicio: string;
  fechaFin: string;
  horaInicio: string;
  horaFin: string;
  notas: string;
  precioTotal: number;
  clienteNombre: string;
  proveedorNombre: string;
  proveedorEmpresa: string;
  servicioNombre: string;
  mascotaNombre: string;
  estadoReservaNombre: string;
  estadoReservaColor: string;
}

export interface ReservaRequest {
  clienteId: number;
  proveedorId?: number;
  servicioId: number;
  mascotaId: number;
  fechaReserva: string;
  fechaInicio: string;
  fechaFin?: string;
  horaInicio: string;
  horaFin: string;
  notas?: string;
  precioTotal?: number;
  registroVacunacionId?: number;
  modalidadEntrega?: string;
  latitud?: number;
  longitud?: number;
  direccionReferencia?: string;
}

export interface SlotDisponible {
  horaInicio: string;
  horaFin: string;
}

export interface ModalidadInfo {
  id: number;
  modalidad: string;
  costoAdicional: number;
}

export interface DisponibilidadSlots {
  fecha: string;
  slots: SlotDisponible[];
  requiereCertificado: boolean;
  modalidades?: ModalidadInfo[];
}

@Injectable({ providedIn: 'root' })
export class ReservaService {
  constructor(private api: ApiService) {}

  create(data: ReservaRequest): Observable<ApiResponse<Reserva>> {
    return this.api.post('/reservas', data);
  }

  getById(id: number): Observable<ApiResponse<Reserva>> {
    return this.api.get(`/reservas/${id}`);
  }

  getAll(params?: any): Observable<ApiResponse<PagedResponse<Reserva>>> {
    return this.api.getPaged('/reservas', params?.page, params?.size, params);
  }

  getByCliente(clienteId: number): Observable<ApiResponse<PagedResponse<Reserva>>> {
    return this.api.getPaged('/reservas', 0, 50, { clienteId });
  }

  getByMascota(mascotaId: number): Observable<ApiResponse<PagedResponse<Reserva>>> {
    return this.api.getPaged('/reservas', 0, 50, { mascotaId });
  }

  getSlots(proveedorId: number, servicioId: number, desde?: string, hasta?: string, excluirReservaId?: number): Observable<ApiResponse<DisponibilidadSlots[]>> {
    return this.api.get('/reservas/slots', { proveedorId, servicioId, desde, hasta, excluirReservaId });
  }

  update(id: number, data: ReservaRequest): Observable<ApiResponse<Reserva>> {
    return this.api.put(`/reservas/${id}`, data);
  }

  cancelar(id: number, motivo?: string): Observable<ApiResponse<Reserva>> {
    return this.api.put(`/reservas/${id}/cancelar`, { motivo: motivo || '' });
  }
}

File: services/reservation-service/src/main/java/bo/capital/tec/pet/modules/reserva/service/ReservaService.java
package bo.capital.tec.pet.modules.reserva.service;

import bo.capital.tec.pet.common.api.PagedResponse;
import bo.capital.tec.pet.modules.reserva.command.CancelarReservaCommand;
import bo.capital.tec.pet.modules.reserva.command.ConfirmarReservaCommand;
import bo.capital.tec.pet.modules.reserva.command.RechazarReservaCommand;
import bo.capital.tec.pet.modules.reserva.dto.DisponibilidadSlotsDTO;
import bo.capital.tec.pet.modules.reserva.dto.ReservaRequestDTO;
import bo.capital.tec.pet.modules.reserva.dto.ReservaResponseDTO;
import bo.capital.tec.pet.modules.reserva.dto.ReservaSummaryDTO;

import java.time.LocalDate;
import java.util.List;

public interface ReservaService {

    ReservaResponseDTO create(ReservaRequestDTO dto);

    List<DisponibilidadSlotsDTO> getSlots(Long proveedorId, Long servicioId,
                                          LocalDate desde, LocalDate hasta, Long excluirReservaId);

    ReservaResponseDTO getById(Long id);

    PagedResponse<ReservaSummaryDTO> getAll(Long clienteId, Long mascotaId, Long servicioId,
                                            String estado, LocalDate fechaDesde, LocalDate fechaHasta,
                                            int page, int size);

    PagedResponse<ReservaSummaryDTO> getByClienteId(Long clienteId, int page, int size);

    PagedResponse<ReservaSummaryDTO> getByProveedorId(Long proveedorId, int page, int size);

    ReservaResponseDTO update(Long id, ReservaRequestDTO dto);

    void delete(Long id);

    ReservaResponseDTO cancelar(Long id, String motivo);

    ReservaResponseDTO aplicarAceptacion(ConfirmarReservaCommand comando);

    ReservaResponseDTO aplicarRechazo(RechazarReservaCommand comando);

    ReservaResponseDTO compensarPorPagoFallido(CancelarReservaCommand comando);
}
